### DISCUSSION TITLE
Is there a way to execute a stand-alone expression?

### DISCUSSION THREAD

**Discussion by Arshia001** [sebastienros/jint#1188] [Q&A]
The engine only accepts complete scripts. Also, the constructor for NodeList is internal, meaning you can't create new scripts in user code. Is there a way to evaluate expression nodes directly? This is useful for implementing e.g. simple one-line calculations in a system's configuration, while disallowing complex scripts, like so:

```yaml
some-config:
  some-item:
    preprocess: x + 12 * 8 - 1
```

One would use `JSParser.ParseExpression` to parse that config entry as an expression, provide some value for `x`, and evaluate it.

**Comment by lahma** [comment]
I think your best option is to create an `AstVisitor` that validates the input to contain only allowed elements. Parse the script, validate the AST (`Script` instance) , if all OK, then let the string run trough `engine.Evaluate("x + 12 * 8 - 1")`.

**Comment by Arshia001** [comment]
So I guess it's currently not possible then. I don't mind using a validated script, but I'm pretty sure the ability to evaluate individual statements and expressions would prove useful sometime, somewhere. If it's acceptable in theory, I can look into adding the feature.

Also, why run it as a string? You already have a validated and parsed AST. Wouldn't it be better (performance-wise) to run the AST directly?

**Reply by lahma** [reply]
Yes,  run the script instance directly so no re-parsing overhead.

I don't think that running only an expression is an easy feat as JS initialization etc is all built based on analyzing script structure. Validated script is basically running only expressions, without changes to engine logic.

**Reply by Arshia001** [reply]
Actually, I think it would (in its most basic form) be as simple as:

```csharp
Execute(Expression exp) => Execute(new Script(new NodeList(new[] { new ExpressionStatement(exp) })));

Execute(Statement stmt) => Execute(new Script(new NodeList(new[] { stmt })));
```

**Reply by lahma** [reply]
Then one could ask for ability to introduce variables too, then we're supporting statement list.. I really think that validating the script instance is easiest for such specific use cases 🙂
